components/mpas-ocean/src/shared/mpas_ocn_vel_self_attraction_loading.F

```diff
@@ -0,0 +1,2457 @@
+           do iCell = 1,nCellsAll
+               d = distanceToCoast(iCell)
+               coastalSmoothingFactor(iCell) = 0.5_RKIND*(tanh((d - trans_start - 0.5_RKIND * trans_width) / (0.2_RKIND *trans_width)) + 1.0_RKIND)
+           enddo
+        else
+           coastalSmoothingFactor(:) = 1.0_RKIND
+        endif
+
+        ! Setup Alarm for SAL
+        alarmTime = mpas_get_clock_time(domain % clock, MPAS_START_TIME, ierr=err)
+        call mpas_set_timeInterval(alarmTimeStep, timeString = &
+              config_self_attraction_loading_compute_interval, ierr = err)
+        if (config_do_restart) then
+          call mpas_add_clock_alarm(domain % clock, 'salComputeAlarm', alarmTime + alarmTimeStep, &
+                alarmTimeInterval=alarmTimeStep, ierr=err)
+        else
+          call mpas_add_clock_alarm(domain % clock, 'salComputeAlarm', alarmTime, &
+                alarmTimeInterval=alarmTimeStep, ierr=err)
+        endif
+
+        if (.not. config_use_parallel_self_attraction_loading) then
+
+#ifndef USE_SHTNS
+            call mpas_log_write('MPAS-Ocean must be compiled with USE_SHTNS=true for serial SAL option.', MPAS_LOG_CRIT)
+#endif
+
+            call mpas_pool_get_array(meshPool, 'indexToCellID', indexToCellID)
+
+            ! Begin MPI portion
+            call MPI_COMM_RANK( domain % dminfo % comm, curProc, err)
+            call MPI_COMM_SIZE( domain % dminfo % comm, nProcs, err)
+
+            if (curProc.eq.0) then
+                allocate(nCellsPerProc(nProcs))
+                allocate(nCellsDisplacement(nProcs))
+            endif
+
+            ! Gather nCellsOwned
+            call MPI_GATHER( nCellsOwned, 1, MPI_INTEGER, nCellsPerProc, 1, MPI_INTEGER, &
+                             0, domain % dminfo % comm, err)
+
+            ! Set Displacement variable for GATHERV command
+            if (curProc.eq.0) then
+                nCellsGlobal = sum(nCellsPerProc)
+                allocate(indexToCellIDGathered(nCellsGlobal))
+                nCellsDisplacement(1) = 0
+                if (nProcs > 1) then
+                    do iProc=2,nProcs
+                        nCellsDisplacement(iProc) = nCellsDisplacement(iProc-1) + nCellsPerProc(iProc-1)
+                    enddo
+                endif
+            endif
+
+            ! Gather indexToCellID
+            call MPI_GATHERV( indexToCellID, nCellsOwned, MPI_INTEGER, indexToCellIDGathered, &
+                    nCellsPerProc, nCellsDisplacement, MPI_INTEGER, 0, domain % dminfo % comm, err)
+
+            ! SAL calculations only need to be initialized on process 1
+            if (curProc.eq.0) then
+                ! Set variables for file locations
+                mpasToGridFile = trim(config_mpas_to_grid_weights_file)
+                gridToMpasFile = trim(config_grid_to_mpas_weights_file)
+
+                ! Open netcdf weights files
+                call check( nf90_open(path = mpasToGridFile, mode = nf90_nowrite, ncid = toNcId) ,err)
```

Comment:
Throughout this section, and elsewhere in this module, you should use a temporary error variable for the check, and then ior it with the existing err variable.  Otherwise an error early in the routine could be overwritten by a success later in the routine.  The typical way we do this is like:
```
call check( nf90_open(path = mpasToGridFile, mode = nf90_nowrite, ncid = toNcId), err_tmp)
err = ior(err, err_tmp)
```

You could also modify the `check` routine to do that for you.  If so, you'll still want to make similar adjustments in the various MPI_* routines (like MPI_SCATTERV).

(I noticed this after reviewing code Holly copied for her sea level model PR into MALI: https://github.com/MALI-Dev/E3SM/pull/21/files)

Reply:
@sbrus89 could you add this `ior` line after each time an error status is returned from a subroutine? Thanks.

Reply:
@mark-petersen, yes I'll do that. @knbarton has a PR into my fork adding these lines.